reductstore/src/storage/block_manager/block.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright 2024 ReductSoftware UG
+// Licensed under the Business Source License 1.1
+
+use crate::storage::proto::{ts_to_us, us_to_ts, Block as BlockProto, Record};
+use prost::Message;
+use reduct_base::error::{ErrorCode, ReductError};
+use std::cmp::min;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+
+#[derive(Clone, Debug, PartialEq)]
+pub(in crate::storage) struct Block {
+    block_id: u64,
+    size: u64,
+    record_count: u64,
+    metadata_size: u64,
+    record_index: BTreeMap<u64, Record>,
+}
+
+impl From<BlockProto> for Block {
+    fn from(inner: BlockProto) -> Self {
+        let mut record_index = BTreeMap::new();
+        for record in inner.records {
+            record_index.insert(ts_to_us(&record.timestamp.unwrap()), record);
+        }
+
+        Block {
+            block_id: ts_to_us(&inner.begin_time.unwrap()),
+            size: inner.size,
+            record_count: inner.record_count,
+            metadata_size: inner.metadata_size,
+            record_index,
+        }
+    }
+}
+
+impl From<Block> for BlockProto {
+    fn from(block: Block) -> Self {
+        let latest_record_time = us_to_ts(&block.latest_record_time());
+        let mut records = Vec::new();
+        for (_, record) in block.record_index {
+            records.push(record);
+        }
+
+        BlockProto {
+            begin_time: Some(us_to_ts(&block.block_id)),
+            latest_record_time: Some(latest_record_time),
+            size: block.size,
+            record_count: block.record_count,
+            metadata_size: block.metadata_size,
+            records,
+            invalid: false,
+        }
+    }
+}
+
+impl Block {
+    pub fn new(block_id: u64) -> Self {
+        Block {
+            block_id,
+            size: 0,
+            record_count: 0,
+            metadata_size: 0,
+            record_index: BTreeMap::new(),
+        }
+    }
+
+    pub fn insert_or_update_record(&mut self, record: Record) {
+        match self
+            .record_index
+            .entry(ts_to_us(&record.timestamp.unwrap()))
+        {
+            Entry::Occupied(mut entry) => {
+                let existing_record = entry.get_mut();
+                self.size -= existing_record.end - existing_record.begin;
+                self.metadata_size -= min(self.metadata_size, existing_record.encoded_len() as u64);
```

Comment:
does it need a warning if metadata_size is smaller than existing_record.encoded_len()?

Reply:
This is somehow normal. The `self.metada_size` is updated with the real size of the block descriptor when it is serialized to disk. I'm counting a rough estimate here, because I don't want to screw up the performance by serializing the descriptor for every new record to know the real size.